### ISSUE TITLE
Android build with android_stl=no fails with NDK r20

### ISSUE THREAD

**Issue by romikusumabakti** [godotengine/godot#30688]
**Godot version:**
Godot 2.1.6 stable

**OS/device including version:**
Windows 10 64-bit, Python 2.7.16, SCons 3.0.5, Android NDK r20

**Issue description:**
Compile error as shown on command line log below. It works fine with NDK r17 on Windows 10 64-bit.

**Steps to reproduce:**
scons platform=android tools=no

**Minimal reproduction project:**
```
[ 98%] Linking Shared Library ==> bin\libgodot.android.opt.armv7.neon.so
[100%] progress_finish(["progress_finish"], [])
[100%] =====
D:\Android\sdk\ndk-bundle/toolchains/arm-linux-androideabi-4.9/prebuilt/windows-x86_64/lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin\ld: error: cannot find -lc++
platform\android\os_android.os:os_android.cpp:function OS_Android::~OS_Android(): error: undefined reference to 'operator delete(void*)'
platform\android\file_access_android.os:file_access_android.cpp:function FileAccessAndroid::~FileAccessAndroid(): error: undefined reference to 'operator delete(void*)'
platform\android\audio_driver_opensl.os:audio_driver_opensl.cpp:function AudioDriverOpenSL::~AudioDriverOpenSL(): error: undefined reference to 'operator delete(void*)'
platform\android\file_access_jandroid.os:file_access_jandroid.cpp:function FileAccessJAndroid::~FileAccessJAndroid(): error: undefined reference to 'operator delete(void*)'
platform\android\ifaddrs_android.os:ifaddrs_android.cpp:function getifaddrs(ifaddrs**): error: undefined reference to 'operator new(unsigned int)'
platform\android\ifaddrs_android.os:ifaddrs_android.cpp:function getifaddrs(ifaddrs**): error: undefined reference to 'operator new[](unsigned int)'
platform\android\ifaddrs_android.os:ifaddrs_android.cpp:function getifaddrs(ifaddrs**): error: undefined reference to 'operator new(unsigned int)'
platform\android\ifaddrs_android.os:ifaddrs_android.cpp:function getifaddrs(ifaddrs**): error: undefined reference to 'operator new(unsigned int)'
platform\android\ifaddrs_android.os:ifaddrs_android.cpp:function getifaddrs(ifaddrs**): error: undefined reference to 'operator new(unsigned int)'
platform\android\ifaddrs_android.os:ifaddrs_android.cpp:function getifaddrs(ifaddrs**): error: undefined reference to 'operator delete[](void*)'
platform\android\ifaddrs_android.os:ifaddrs_android.cpp:function getifaddrs(ifaddrs**): error: undefined reference to 'operator delete[](void*)'
platform\android\ifaddrs_android.os:ifaddrs_android.cpp:function getifaddrs(ifaddrs**): error: undefined reference to 'operator delete[](void*)'
platform\android\ifaddrs_android.os:ifaddrs_android.cpp:function freeifaddrs(ifaddrs*): error: undefined reference to 'operator delete[](void*)'
modules\libmodules.android.opt.armv7.neon.a(register_types.android.opt.armv7.neon.o):register_types.cpp:vtable for AudioStream: error: undefined reference to '__cxa_pure_virtual'
modules\libmodules.android.opt.armv7.neon.a(stream_peer_openssl.android.opt.armv7.neon.o):stream_peer_openssl.cpp:vtable for StreamPeerSSL: error: undefined reference to '__cxa_pure_virtual'
modules\libmodules.android.opt.armv7.neon.a(stream_peer_openssl.android.opt.armv7.neon.o):stream_peer_openssl.cpp:vtable for StreamPeerSSL: error: undefined reference to '__cxa_pure_virtual'
modules\libmodules.android.opt.armv7.neon.a(stream_peer_openssl.android.opt.armv7.neon.o):stream_peer_openssl.cpp:vtable for StreamPeerSSL: error: undefined reference to '__cxa_pure_virtual'
modules\libmodules.android.opt.armv7.neon.a(texture_loader_pvr.android.opt.armv7.neon.o):texture_loader_pvr.cpp:function _compress_pvrtc4(Image*): error: undefined reference to 'operator new[](unsigned int)'
drivers\libdrivers.android.opt.armv7.neon.a(rasterizer_gles2.android.opt.armv7.neon.o):rasterizer_gles2.cpp:function RasterizerGLES2::_render(RasterizerGLES2::Geometry const*, RasterizerGLES2::Material const*, RasterizerGLES2::Skeleton const*, RasterizerGLES2::GeometryOwner const*, Transform const&): error: undefined reference to '__cxa_guard_acquire'
drivers\libdrivers.android.opt.armv7.neon.a(rasterizer_gles2.android.opt.armv7.neon.o):rasterizer_gles2.cpp:function RasterizerGLES2::_render(RasterizerGLES2::Geometry const*, RasterizerGLES2::Material const*, RasterizerGLES2::Skeleton const*, RasterizerGLES2::GeometryOwner const*, Transform const&): error: undefined reference to '__cxa_guard_release'
drivers\libdrivers.android.opt.armv7.neon.a(rasterizer_gles2.android.opt.armv7.neon.o):rasterizer_gles2.cpp:function RasterizerGLES2::_render(RasterizerGLES2::Geometry const*, RasterizerGLES2::Material const*, RasterizerGLES2::Skeleton const*, RasterizerGLES2::GeometryOwner const*, Transform const&): error: undefined reference to '__cxa_guard_acquire'
drivers\libdrivers.android.opt.armv7.neon.a(rasterizer_gles2.android.opt.armv7.neon.o):rasterizer_gles2.cpp:function RasterizerGLES2::_render(RasterizerGLES2::Geometry const*, RasterizerGLES2::Material const*, RasterizerGLES2::Skeleton const*, RasterizerGLES2::GeometryOwner const*, Transform const&): error: undefined reference to '__cxa_guard_release'
drivers\libdrivers.android.opt.armv7.neon.a(rasterizer_gles2.android.opt.armv7.neon.o):rasterizer_gles2.cpp:function RasterizerGLES2::_render(RasterizerGLES2::Geometry const*, RasterizerGLES2::Material const*, RasterizerGLES2::Skeleton const*, RasterizerGLES2::GeometryOwner const*, Transform const&): error: undefined reference to '__cxa_guard_acquire'
drivers\libdrivers.android.opt.armv7.neon.a(rasterizer_gles2.android.opt.armv7.neon.o):rasterizer_gles2.cpp:function RasterizerGLES2::_render(RasterizerGLES2::Geometry const*, RasterizerGLES2::Material const*, RasterizerGLES2::Skeleton const*, RasterizerGLES2::GeometryOwner const*, Transform const&): error: undefined reference to '__cxa_guard_release'
drivers\libdrivers.android.opt.armv7.neon.a(rasterizer_gles2.android.opt.armv7.neon.o):rasterizer_gles2.cpp:function RasterizerGLES2::_render(RasterizerGLES2::Geometry const*, RasterizerGLES2::Material const*, RasterizerGLES2::Skeleton const*, RasterizerGLES2::GeometryOwner const*, Transform const&): error: undefined reference to '__cxa_guard_acquire'
drivers\libdrivers.android.opt.armv7.neon.a(rasterizer_gles2.android.opt.armv7.neon.o):rasterizer_gles2.cpp:function RasterizerGLES2::_render(RasterizerGLES2::Geometry const*, RasterizerGLES2::Material const*, RasterizerGLES2::Skeleton const*, RasterizerGLES2::GeometryOwner const*, Transform const&): error: undefined reference to '__cxa_guard_release'
core\libcore.android.opt.armv7.neon.a(triangulator.android.opt.armv7.neon.o):triangulator.cpp:function TriangulatorPoly::Init(long): error: undefined reference to 'operator new[](unsigned int)'
core\libcore.android.opt.armv7.neon.a(triangulator.android.opt.armv7.neon.o):triangulator.cpp:function TriangulatorPoly::operator=(TriangulatorPoly const&): error: undefined reference to 'operator new[](unsigned int)'
clang++: error: linker command failed with exit code 1 (use -v to see invocation)

=====
scons: *** [bin\libgodot.android.opt.armv7.neon.so] Error 1
scons: building terminated because of errors.
```

**Comment by welldiniz** [comment]
I am also having the same problem, need to resolve fast or I will lose my applications in google play.
Can anyone help?

**Comment by akien-mga** [comment]
You can use NDK r18 which works fine in the meantime.

**Comment by akien-mga** [comment]
I can confirm a similar build issue on Linux with NDK r19c (and I assume r20 too):
```
Linking Shared Library ==> bin/libgodot.android.debug.armv7.neon.so
/home/akien/Projects/godot/godot-2.1/drivers/unix/os_unix.cpp:159: error: undefined reference to 'sigemptyset'
/home/akien/Projects/godot/godot-2.1/drivers/unix/os_unix.cpp:184: error: undefined reference to 'stdout'
/home/akien/Projects/godot/godot-2.1/drivers/unix/os_unix.cpp:184: error: undefined reference to 'stderr'
/home/akien/Projects/godot/godot-2.1/drivers/unix/os_unix.cpp:184: error: undefined reference to 'stderr'
/home/akien/Projects/godot/godot-2.1/drivers/unix/os_unix.cpp:196: error: undefined reference to 'stderr'
/home/akien/Projects/godot/godot-2.1/drivers/unix/os_unix.cpp:236: error: undefined reference to 'stdin'
/home/akien/Projects/godot/godot-2.1/drivers/unix/os_unix.cpp:236: error: undefined reference to 'stdin'
/home/akien/Projects/godot/godot-2.1/drivers/unix/os_unix.cpp:236: error: undefined reference to 'stdin'
/home/akien/Projects/godot/godot-2.1/drivers/unix/os_unix.cpp:203: error: undefined reference to '__FD_SET_chk'
/home/akien/Projects/godot/godot-2.1/drivers/unix/os_unix.cpp:411: error: undefined reference to 'stderr'
/home/akien/Projects/godot/godot-2.1/./drivers/unix/socket_helpers.h:72: error: undefined reference to 'in6addr_any'
/home/akien/Projects/godot/godot-2.1/./drivers/unix/socket_helpers.h:72: error: undefined reference to 'in6addr_any'
/home/akien/Projects/godot/godot-2.1/thirdparty/libvorbis/psy.c:126: error: undefined reference to 'abs'
/home/akien/Projects/godot/godot-2.1/thirdparty/libvorbis/psy.c:1136: error: undefined reference to 'abs'
/home/akien/Projects/godot/godot-2.1/thirdparty/libvorbis/psy.c:1136: error: undefined reference to 'abs'
/home/akien/Projects/godot/godot-2.1/thirdparty/libvorbis/psy.c:1144: error: undefined reference to 'abs'
/home/akien/Projects/godot/godot-2.1/core/io/pck_packer.cpp:179: error: undefined reference to 'stdout'
/home/akien/Projects/godot/godot-2.1/modules/openssl/stream_peer_openssl.cpp:177: error: undefined reference to 'stdout'
/home/akien/Projects/godot/godot-2.1/modules/openssl/stream_peer_openssl.cpp:426: error: undefined reference to 'stdout'
/home/akien/Projects/godot/godot-2.1/thirdparty/rg-etc1/rg_etc1.cpp:1806: error: undefined reference to 'labs'
/home/akien/Projects/godot/godot-2.1/thirdparty/rg-etc1/rg_etc1.cpp:1820: error: undefined reference to 'labs'
/home/akien/Projects/godot/godot-2.1/thirdparty/rg-etc1/rg_etc1.cpp:1912: error: undefined reference to 'labs'
/home/akien/Projects/godot/godot-2.1/thirdparty/opus/stream.c:102: error: undefined reference to 'fseeko64'
/home/akien/Projects/godot/godot-2.1/thirdparty/opus/stream.c:117: error: undefined reference to 'ftello64'
/home/akien/Projects/godot/godot-2.1/thirdparty/openssl/crypto/ui/ui_openssl.c:497: error: undefined reference to 'tcgetattr'
/home/akien/Projects/godot/godot-2.1/thirdparty/openssl/crypto/ui/ui_openssl.c:557: error: undefined reference to 'stdin'
/home/akien/Projects/godot/godot-2.1/thirdparty/openssl/crypto/ui/ui_openssl.c:665: error: undefined reference to 'signal'
/home/akien/Projects/godot/godot-2.1/thirdparty/openssl/crypto/ui/ui_openssl.c:567: error: undefined reference to 'tcsetattr'
/home/akien/Projects/godot/godot-2.1/thirdparty/openssl/crypto/ui/ui_openssl.c:588: error: undefined reference to 'tcsetattr'
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
```

The 3.0 branch builds fine on the other hand.

**Comment by akien-mga** [comment]
CC @rdhafidh and @allkhor who last worked on the relevant code in `detect.py`.

I just cherry-picked #21510 by @milkowski as a fix up to #20507 which had been included in the 3.0 and master/3.1 branches but not 2.1. It doesn't fix the issue though, but I figured it would be good to include the fix nevertheless.

**Comment by allkhor** [comment]
I compiled with docker and release game used NDK r19c - Pkg.Revision = 19.2.5345600 without any issues. Tried r20 and get same errors. As workaround we can use android_stl=yes, but this increase final size about 4mb. I can't compile master with "android_stl=no", because i have error:
```
servers/visual/visual_server_scene.cpp:35:10: fatal error: 'new' file not found
#include <new>
         ^~~~~
1 error generated.
```
 
After this commit: 3e4dab538656c0fb71 not possible compile without android_stl.
Godot 3.0.6 with "android_stl=no" provide same error.

Also related https://github.com/android-ndk/ndk/issues/951https://github.com/android-ndk/ndk/issues/951

**Comment by akien-mga** [comment]
Fixed by #30758 (at least on r19c - I'll test r20 tomorrow to confirm).

**Comment by allkhor** [comment]
I think bug fixed only for r19,  r20 don't compiled.
Godot 3.0.6 with "android_stl=no" not compiled too.

**Comment by akien-mga** [comment]
>  Godot 3.0.6 with "android_stl=no" not compile too.

We use various thirdparty libraries which depend on the STL, so I don't think `android_stl=no` is meant to work anymore out of the box. It was defaulted to `yes` due to those libraries. It could be made to work again but that would imply disabling the modules using those thirdparty libraries if `android_stl=no`, and enclose changes in https://github.com/godotengine/godot/commit/3e4dab538656c0fb71ac0e2535a1e51c2b9ebe06 in some `#ifdef`.

**Comment by allkhor** [comment]
I agree, but I compile Godot 3.0.6 with disabled modules depend on the STL and get error when compiler try to link. Maybe this link can helps: https://android.googlesource.com/platform/ndk/+/ndk-release-r20/docs/BuildSystemMaintainers.md

**Comment by akien-mga** [comment]
Thanks, the `r19` build issue was separate then, I'll retitle this issue to focus on the `android_stl=no` build. `r20` with `android_stl=yes` does seem to build fine.

The problem likely comes from this change between [r19 docs](https://android.googlesource.com/platform/ndk/+/ndk-release-r19/docs/BuildSystemMaintainers.md#STL) and [r20 docs](https://android.googlesource.com/platform/ndk/+/ndk-release-r20/docs/BuildSystemMaintainers.md#STL):

> - r19: libc++: [...] To use this STL, use the -stdlib=libc++ flag.
> - r20: libc++: [...] This STL is used by default.

Following this when using `android_stl=no` might be sufficient to fix it:

> To avoid using the STL at all, pass -nostdinc++ when compiling and -nostdlib++ when linking.

I assume `new` and `delete` then still work because we rely on the "System STL", which they plan to remove eventually. So in the future it likely won't be possible to build without STL support, unless we drop all usage of `new` and `delete` (which are also used in our own Android platform code).

**Comment by akien-mga** [comment]
I was confused above, actually both `r19c` and `r20` fail building with `android_stl=no`.

I'll try older NDKs to see if it did work previously, or if we simply lost support for building without STL long ago and we should just drop the option.

**Comment by akien-mga** [comment]
> I was confused above, actually both `r19c` and `r20` fail building with `android_stl=no`.

Aaaand I got confused again. `r19c` does build fine with both `android_stl=yes` and `android_stl=no`. `r20` builds fine with `android_stl=yes` and fails with `android_stl=no` (on 2.1).
 
So back to looking for what changed in `r20` to break the option. I guess maybe they did drop the "System STL" that they previously mentioned as deprecated? It's not reflected in the docs though.

> I'll try older NDKs to see if it did work previously, or if we simply lost support for building without STL long ago and we should just drop the option.

Just tried `r14b` which also builds fine with both options.

**Comment by akien-mga** [comment]
Fixed for 2.1 in #30963 and for 3.0 in #30965. For 3.1+, I'll drop the option as it stops making sense will all the C++ dependencies we have.
